### ISSUE TITLE
Simple tracker example code.

### ISSUE THREAD

**Issue by nemanja-rakicevic** [ValveSoftware/openvr#250]
Hello, has anyone tried to make a simple .cpp code to extract and print the controller/hmd pose coordinates in the terminal (possibly in Linux)?
I am having problems understanding the API functions needed to perform just this task.
Thanks!

**Comment by echuber2** [comment]
What parts are you confused about? The hellovr example has functions checking the orientations and manipulating the matrices.

**Comment by nemanja-rakicevic** [comment]
Exactly, but it's ~2000 lines of not very well documented code. I was wondering if there are some simpler examples, which don't necessarily include graphics as well, but just reading position&orientation and printing it. Maybe someone already made this, so I save time since I plan to forward this to robot's actuators and get a feed from Kinect to display on the HMD.

P.S. - Is there a good python wrapper maybe? I tried https://github.com/cmbruns/pyopenvr but it's not installing properly

**Comment by echuber2** [comment]
Focus on the functions in hellovr that have "matrix" in the name. I'm
interested in writing a more minimal example like you've described for
accessibility purposes but I doubt I can find the time this week. The
mathematics of the orientation matrices have already been discussed here
and in other places (Steam forums, for example).

On Wed, Aug 31, 2016 at 8:23 AM, nemanja-rakicevic <notifications@github.com

**Comment by echuber2** [comment]
Focus on the functions in hellovr that have "matrix" in the name. I'm interested in writing a more minimal example like you've described for accessibility purposes but I doubt I can find the time this week. The mathematics of the orientation matrices have already been discussed here and in other places (Steam forums, for example).

**Comment by nemanja-rakicevic** [comment]
Yes, I'm looking into it at the moment.
If you manage to do something like this I would be extremely grateful (as well as many future beginners :)

I guess there shouldn't be any problems compiling the .cpp files on Linux, right?

**Comment by profiler-bg** [comment]
Hi, what exactly do you want to print out? Any example format you want to see? I might be able to do it today or tomorrow.

**Comment by nemanja-rakicevic** [comment]
Hello, I managed to find a ros package which prints the frame poses https://github.com/robosavvy/vive_ros and I managed to edit it a bit to publish the button states. However, I was not able to execute the vibration of the controllers.

Moreover, the most important point I'm missing now is how to render projections based on some 3D model.. the example does it only through openGL it seems. It would be very useful if I could find some documentation on rendering.

**Comment by echuber2** [comment]
What do you mean by rendering projections exactly? If not OpenGL, are you meaning to use software rendering? If you just need a primer on the math, you can refer to sites like this one. http://www.opengl-tutorial.org/beginners-tutorials/tutorial-3-matrices/

**Comment by nemanja-rakicevic** [comment]
No, the theory behind - the geometry/math part is fine.
So based on the hellovr example, as I understand, to get the images you see on the HMD, you need to create a 3D object (in this case cubes using vertices etc, in openGL) and then project them based on the HMD's pose.

What I am looking for is a way to import a 3D model, or a 3D point cloud (from Kinect, SLAM map etc) which is updated regularly (e.g. kinect video feed) and project it directly to the HMD. Maybe as you said I would need additional software. (on Linux)
Thanks!

**Comment by echuber2** [comment]
The way it works is rather like a conventional game engine with a camera, only in this case there are two cameras (one for each eye) and the projection matrix (based on the viewing frustum) is basically set in stone and governed by the OpenVR stack for you, based on the HMD. The hellovr example shows how this is done; it also loads some objects with the CGLRenderModel class, although the usage is not so clear. Basically you might want to look into using a game engine with OpenVR for ease of use, if you're not comfortable.

If you want to use a seated configuration rather than the room-space features, you should look at the function ResetSeatedZeroPose() and then you can call GetDeviceToAbsoluteTrackingPose passing vr::TrackingUniverseSeated. (See the OpenVR documentation on GitHub.)

The main challenge in this is deciding on your frame of reference. In seated configuration, the user's head becomes the origin, and after the ResetSeatedZeroPose() the head is tracked through it. So you can put something in front of the user's face by placing it at (0,0,-1), or so forth.
